### ISSUE TITLE
Add stable log1mexp = log(1 - exp(-|x|)) function

### ISSUE THREAD

**Issue by simonepri** [pytorch/pytorch#39242]
## 🚀 Feature
Add the implementation for a numerically stable `log(1 - exp(-|x|))` function.

See https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf

## Motivation

The function can be implemented using other pytorch functions but it might be tricky to get it right.

Some implementations found on GitHub:

[@n3net/blob/5d5883a/src_denoising/models/non_local.py#L94-L108](https://github.com/wouterkool/estimating-gradients-without-replacement/blob/9d8bf8b/bernoulli/gumbel.py#L7-L11)

[@n3net/blob/5d5883a/src_denoising/models/non_local.py#L94-L108](https://github.com/visinf/n3net/blob/5d5883a/src_denoising/models/non_local.py#L94-L108)

**Comment by mruberry** [comment]
Hey @simonepri, thanks for the suggestion. Would you be interested in submitting a PR for it?

Follow-up question, how numerically stable are PyTorch's log1p and expm1 implementations?

**Comment by cmpute** [comment]
Thanks for the links provided! This function is also useful in my work!

**Comment by Balandat** [comment]
If anyone still wants to add this (courtesy of @SebastianAment): 
```
def log1mexp(x: Tensor) -> Tensor:
    """Numerically accurate evaluation of log(1 - exp(x)) for x < 0.
    See [Maechler2012accurate]_ for details.
    """
    mask = -math.log(2) < x  # x < 0
    return torch.where(
        mask,
        (-x.expm1()).log(),
        (-x.exp()).log1p(),
    )
```

[Maechler2012accurate] M. Mächler. Accurately Computing log (1 - exp (-| a|)). Assessed by the Rmpfr package. Technical report, 2012. https://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf

**Comment by abcde2342** [comment]
Please take a look at my Taylor-series-style patch for log1mexp in the dangerous region.

```
from __future__ import annotations

import math

from typing import Any

import mpmath as mp
import torch
import torch.nn.functional as F
import numpy as np

mp.mp.dps = 100

CASE_TITLE = 'Add numerically stable log1mexp = log(1 - exp(-|x|)) function'

CASE_FUNCTION = 'log1mexp = log(1 - exp(x)), x < 0'

CASE_ID = 'pytorch/pytorch#39242'

FINITE_DIFF_FIXED_STEP = mp.mpf("1e-5")

def relerr(v, ref):
    is_complex = isinstance(v, complex) or isinstance(ref, (complex, mp.ctx_mp_python.mpc))
    v = mp.mpc(v) if is_complex else mp.mpf(v)
    ref = mp.mpc(ref) if is_complex else mp.mpf(ref)
    if abs(ref) == 0:
        return float(abs(v - ref))
    return float(abs((v - ref) / ref))

def orders(before, after):
    if not math.isfinite(before) or before <= 0:
        return 'inf' if after == 0 or math.isfinite(after) else 'nan'
    if after == 0:
        return 'inf'
    if not math.isfinite(after):
        return '-inf'
    return f'{math.log10(before / after):.2f}'

def log1p_taylor_scalar(s, terms=8):
    total = 0.0
    power = 1.0
    for n in range(1, terms + 1):
        power *= s
        total += (-1.0) ** (n + 1) * power / n
    return total

def log1mexp_naive(x):
    return math.log(1.0 - math.exp(x))

def expm1_taylor_scalar(x, terms=12):
    total = 0.0
    power = 1.0
    fact = 1.0
    for n in range(1, terms + 1):
        power *= x
        fact *= n
        total += power / fact
    return total

def log1mexp_taylor(x, terms=12):
    if x > -math.log(2.0):
        return math.log(-expm1_taylor_scalar(x, terms=terms))
    t = -math.exp(x)
    return log1p_taylor_scalar(t, terms=terms)

def _mp_value(x):
    if isinstance(x, (complex, mp.ctx_mp_python.mpc)):
        return mp.mpc(x)
    return mp.mpf(str(float(x))) if isinstance(x, np.generic) else mp.mpf(str(x))

def _finite_difference_step(x):
    return FINITE_DIFF_FIXED_STEP

def _finite_difference_derivative(f, x):
    x = _mp_value(x)
    fx = f(x)
    h = _finite_difference_step(x)

    def candidate(step):
        return ((f(x + step) - fx) / step, step)
    for step in (h, -h):
        try:
            d, used = candidate(step)
            if mp.isfinite(abs(d)):
                return (d, used, fx)
        except Exception:
            continue
    return (mp.nan, h, fx)

def _condition_from_finite_difference(f, x):
    d, h, fx = _finite_difference_derivative(f, x)
    x = _mp_value(x)
    if abs(fx) == 0:
        value = abs(d)
    else:
        value = abs(x * d / fx)
    return (float(value), h)

def _format_value(value: Any) -> str:
    try:
        if isinstance(value, (mp.ctx_mp_python.mpf, mp.ctx_mp_python.mpc)):
            return mp.nstr(value, 17)
        if isinstance(value, complex):
            return str(value)
        if isinstance(value, np.generic):
            return repr(value.item())
        return repr(value)
    except Exception:
        return str(value)

def case_pytorch_pytorch_39242() -> dict[str, Any]:
    case = 'pytorch/pytorch#39242'
    expressions = ('log(1-exp(x))', 'mpmath log(1-exp(x))', 'log(-(x + x**2/2 + x**3/6 + ...))')
    xval = -1e-16
    ref = mp.log(1 - mp.e ** mp.mpf(str(xval)))
    try:
        bval = log1mexp_naive(xval)
    except ValueError:
        bval = math.nan
    before = relerr(bval, ref) if math.isfinite(bval) else math.inf
    taylor_value = log1mexp_taylor(xval)
    after = relerr(taylor_value, ref)
    condition = _condition_from_finite_difference(lambda xx: mp.log(1 - mp.e ** xx), xval)
    row = {'condition_number': condition[0], 'original_expression': expressions[0], 'oracle_expression': expressions[1], 'taylor_expression': expressions[2], 'original_value': _format_value(bval), 'oracle_value': _format_value(ref), 'taylor_value': _format_value(taylor_value), 'original_relative_error': before, 'taylor_relative_error': after, 'taylor_orders_improved': orders(before, after)}
    return ([row], expressions[2])

def print_case_report() -> None:
    rows, patch_idea = case_pytorch_pytorch_39242()
    print(f"{CASE_ID}: {CASE_TITLE}")
    print(f"  function: {CASE_FUNCTION}")
    print(f"  patch: {patch_idea}")
    for row in rows:
        print("  evidence:")
        print(f"    condition_number={float(row['condition_number']):.6e}")
        print(f"    original_expression={row['original_expression']}")
        print(f"    oracle_expression={row['oracle_expression']}")
        print(f"    taylor_expression={row['taylor_expression']}")
        print(f"    original_value={row['original_value']}")
        print(f"    oracle_value={row['oracle_value']}")
        print(f"    taylor_value={row['taylor_value']}")
        print(f"    original_relative_error={float(row['original_relative_error']):.6e}")
        print(f"    taylor_relative_error={float(row['taylor_relative_error']):.6e}")
        print(f"    taylor_orders_improved={row['taylor_orders_improved']}")

if __name__ == "__main__":
    print_case_report()

```
